request.lisp

```diff
@@ -118,35 +118,41 @@ body using the boundary BOUNDARY."
                    (format stream "Content-Type: ~A" content-type)
                    (crlf) (crlf)
                    ;; use SEND-CONTENT to send file as binary data
-                   (send-content file-source stream)))
+                   (send-content file-source stream :buffer-size buffer-size)))
                 (t (parameter-error
                     "Don't know what to do with name/value pair (~S . ~S) in multipart/form-data body."
                     name value)))
           (crlf)))
       (format stream "--~A--" boundary)
       (crlf))))
 
-(defun %read-body (stream element-type)
+(defun %read-body (stream element-type &key (buffer-size +buffer-size+))
+  "Helper function to read from stream into a buffer of element-type, which is returned."
   ;; On ABCL, a flexi-stream is not a normal stream. This is caused by
   ;; a bug in ABCL which is supposedly quite difficult to fix. More
   ;; details here: http://abcl.org/trac/ticket/377
   #-abcl
   (declare (stream stream))
-  "Helper function to read from stream into a buffer of element-type, which is returned."
-  (let ((buffer (make-array +buffer-size+ :element-type element-type))
-        (result (make-array 0 :element-type element-type :adjustable t)))
-        (loop for index = 0 then (+ index pos)
-           for pos = (read-sequence buffer stream)
-           do (adjust-array result (+ index pos))
-             (replace result buffer :start1 index :end2 pos)
-           while (= pos +buffer-size+))
-        result))
-
-(defun read-body (stream headers textp &key (decode-content t))
+  ;; In 32-bit Clozure-CL the maximum array size limit
+  ;; is (expt 2 24) elements.
+  ;; See: http://ccl.clozure.com/manual/chapter4.7.html
+  ;; Use a list as buffer
+  (progn
+    (when (>= buffer-size array-total-size-limit)
+      (setq buffer-size (- array-total-size-limit 1)))
```

Comment:
Re-bind buffer size instead of using `setq`.  Do so inside of the `loop` using `with` (see below for more of that).